### PR TITLE
Support camera depth texture

### PR DESCRIPTION
**Pull request by GuoLei1990**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

Feat:

- Camera support `depthTextureMode`
- Camera support `pixelViewport`

Playground:https://github.com/galacean/galacean.github.io/pull/764

### CODE REVIEW

**Review by cptbtptpbcptdtptp — CHANGES_REQUESTED** [review]

File: packages/core/src/shader/ShaderPool.ts [hunk 1]
```diff
@@ -29,19 +31,22 @@ import { ShaderPass } from "./ShaderPass";
...
   static init(): void {
-    const shadowCasterPassTags = {
+    const depthOnlyPass = new ShaderPass(depthOnlyVs, depthOnlyFs, {
+      pipelineStage: PipelineStage.DepthOnly
+    });
+    const shadowCasterPass = new ShaderPass(shadowMapVs, shadowMapFs, {
       pipelineStage: PipelineStage.ShadowCaster
-    };
+    });
+    const basePasses = [depthOnlyPass, shadowCasterPass];
```
Comment (zhuxudong) [inline packages/core/src/shader/ShaderPool.ts:40]: use [shadowCasterPass, depthOnlyPass]  to be compatible with previous usage

File: packages/core/src/shaderlib/extra/depthOnly.vs.glsl [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+uniform mat4 camera_VPMat;
+
+
+void main() {
+
+    #include <begin_position_vert>
+    #include <blendShape_vert>
+    #include <skinning_vert>
+
+    gl_Position = renderer_MVPMat * position;
```
Comment (zhuxudong) [inline packages/core/src/shaderlib/extra/depthOnly.vs.glsl:13]: why not use ` #include <position_vert>`

File: packages/core/src/shaderlib/extra/depthOnly.fs.glsl [hunk 3]
```diff
@@ -0,0 +1,3 @@
+void main() {
+  gl_FragColor = vec4(0.0, 0.0, 0.0, 0.0);
```
Comment (zhuxudong) [inline packages/core/src/shaderlib/extra/depthOnly.fs.glsl:2]: delete this line is better

File: packages/core/src/RenderPipeline/CullingResults.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import { Engine } from "../Engine";
+import { RenderQueue } from "./RenderQueue";
+
+/**
+ * @internal
+ * Culling results.
+ */
+export class CullingResults {
```
Comment (zhuxudong) [inline packages/core/src/RenderPipeline/CullingResults.ts:8]: why not called `RendererQueueManager`

File: packages/core/src/RenderPipeline/PipelineUtils.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+   * @param currentTexture - Current texture
+   * @param width - Need texture width
+   * @param height - Need texture height
+   * @param format - Need texture format
+   * @param mipmap - Need texture mipmap
+   * @returns Texture
+   */
+  static recreateTextureIfNeeded(
+    engine: Engine,
+    currentTexture: Texture2D,
```
Comment (zhuxudong) [inline packages/core/src/RenderPipeline/PipelineUtils.ts:20]: currentTexture: Texture2D | void

File: packages/core/src/RenderPipeline/PipelineUtils.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+   * @param width - Need render target width
+   * @param height - Need render target height
+   * @param colorFormat - Need render target color format
+   * @param depthFormat - Need render target depth format
+   * @param mipmap - Need render target mipmap
+   * @returns Render target
+   */
+  static recreateRenderTargetIfNeeded(
+    engine: Engine,
+    currentRenderTarget: RenderTarget,
```
Comment (zhuxudong) [inline packages/core/src/RenderPipeline/PipelineUtils.ts:56]: currentRenderTarget: RenderTarget | void

File: packages/core/src/shadow/PipelinePass.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { Engine } from "../Engine";
```
Comment (zhuxudong) [inline packages/core/src/shadow/PipelinePass.ts:1]: this file should migrate to `core/src/RenderPipeline`

File: packages/core/src/shadow/PipelinePass.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import { CullingResults } from "../RenderPipeline/CullingResults";
+import { RenderContext } from "../RenderPipeline/RenderContext";
+
+/**
+ * PipelinePass is a base class for all pipeline passes.
+ */
+export abstract class PipelinePass {
+  protected _engine: Engine;
+
+  constructor(engine: Engine) {
```
Comment (zhuxudong) [inline packages/core/src/shadow/PipelinePass.ts:11]: I think based on `Camera` is better, `DepthOnlyPass` and `CascadedShadowCasterPass` will be more suitable

File: packages/core/src/shadow/PipelinePass.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    this._engine = engine;
+  }
+
+  /**
+   * Called before rendering a camera, override this method to configure the camera If you need to configure the camera clear flag or render target.
+   * @param context - Rendering context
+   * @param cullingResults - Culling results
+   */
+  abstract onRender(context: RenderContext, cullingResults: CullingResults): void;
```
Comment (zhuxudong) [inline packages/core/src/shadow/PipelinePass.ts:20]: if based on `camera`, `context` and `cullingResults` can be ignored from `camera.engine._renderContext` and `camera.renderPipeline._cullingResults`

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts [hunk 10]
```diff
@@ -179,7 +173,8 @@ export class BasicRenderPipeline {
     cubeFace?: TextureCubeFace,
     mipLevel?: number
   ) {
-    pass.preRender(camera, this._opaqueQueue, this._alphaTestQueue, this._transparentQueue);
+    const cullingResults = this._cullingResults;
```
Comment (zhuxudong) [inline packages/core/src/RenderPipeline/BasicRenderPipeline.ts:176]: const {opaqueQueue, alphaTestQueue, transparentQueue} = cullingResults;

File: packages/core/src/Camera.ts [hunk 11]
```diff
@@ -93,6 +103,8 @@ export class Camera extends Component {
   @deepClone
   private _viewport: Vector4 = new Vector4(0, 0, 1, 1);
   @deepClone
+  private _pixelViewport: Vector4 = new Vector4(0, 0, 0, 0);
```
[the remark this answers is not in the document]
Comment (GuoLei1990) [reply packages/core/src/Camera.ts:106]: `viewport` use `Vector4`, `pixelViewport` use `Rect` is a little strange! But is OK!

File: packages/core/src/RenderPipeline/DepthOnlyPass.ts [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+      height,
+      null,
+      TextureFormat.Depth16,
+      false
+    );
+    const { depthTexture } = renderTarget;
+    depthTexture.wrapModeU = depthTexture.wrapModeV = TextureWrapMode.Clamp;
+    depthTexture.filterMode = TextureFilterMode.Point;
+
+    this._renderTarget = renderTarget;
```
[the remark this answers is not in the document]
Comment (GuoLei1990) [reply packages/core/src/RenderPipeline/DepthOnlyPass.ts:45]: Depth internal render target use custom destroy, to ignore the memory grow when change the resolution

File: packages/core/src/shader/ShaderPool.ts (hunk 1 above)
Comment (GuoLei1990) [reply packages/core/src/shader/ShaderPool.ts:40]: done

File: packages/core/src/shaderlib/extra/depthOnly.vs.glsl (hunk 2 above)
Comment (GuoLei1990) [reply packages/core/src/shaderlib/extra/depthOnly.vs.glsl:13]: done

File: packages/core/src/RenderPipeline/PipelineUtils.ts (hunk 5 above)
Comment (GuoLei1990) [reply packages/core/src/RenderPipeline/PipelineUtils.ts:20]: done with  `currentTexture: Texture2D | null`

File: packages/core/src/RenderPipeline/PipelineUtils.ts (hunk 6 above)
Comment (GuoLei1990) [reply packages/core/src/RenderPipeline/PipelineUtils.ts:60]: done

File: packages/core/src/shadow/PipelinePass.ts (hunk 7 above)
Comment (GuoLei1990) [reply packages/core/src/shadow/PipelinePass.ts:1]: done

File: packages/core/src/shadow/PipelinePass.ts (hunk 8 above)
Comment (GuoLei1990) [reply packages/core/src/shadow/PipelinePass.ts:11]: Different camera use the same Pipeline pass instance, should not bind camera in constrouctor

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts (hunk 10 above)
Comment (GuoLei1990) [reply packages/core/src/RenderPipeline/BasicRenderPipeline.ts:176]: done

**Review by cptbtptpbcptdtptp — APPROVED** [review]
+1

**Review by zhuxudong — APPROVED** [review]

**Pull request merged**